File: src/bumperbot_py_examples/bumperbot_py_examples/simple_tf_kinematics.py
import rclpy
from rclpy.node import Node
from tf2_ros.static_transform_broadcaster import StaticTransformBroadcaster
from tf2_ros.transform_broadcaster import TransformBroadcaster
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from tf2_ros import TransformException

from geometry_msgs.msg import TransformStamped
from bumperbot_msgs.srv import GetTransform

class SimpleTFKinematics(Node):
    def __init__(self):
        super().__init__("simple_tf_kinematics")
        self.x_increment_ = 0.01
        self.last_x_ = 0.0
        self.tf_buffer_ = Buffer()
        self.tf_listener_ = TransformListener(self.tf_buffer_, self)

        self.static_tf_broadcaster = StaticTransformBroadcaster(self)
        self.dynamic_tf_broadcaster = TransformBroadcaster(self)

        self.static_transform_stamped = TransformStamped()
        self.static_transform_stamped_1 = TransformStamped()
        self.dynamic_transform_stamped = TransformStamped()

        self.static_transform_stamped.header.stamp = self.get_clock().now().to_msg()
        self.static_transform_stamped.header.frame_id = "bumperbot_base"
        self.static_transform_stamped.child_frame_id = "bumperbot_top"
        self.static_transform_stamped.transform.translation.x = 0.0
        self.static_transform_stamped.transform.translation.y = 0.0
        self.static_transform_stamped.transform.translation.z = 0.3
        self.static_transform_stamped.transform.rotation.x = 0.0
        self.static_transform_stamped.transform.rotation.y = 0.0
        self.static_transform_stamped.transform.rotation.z = 0.0
        self.static_transform_stamped.transform.rotation.w = 1.0

        self.static_transform_stamped_1.header.stamp = self.get_clock().now().to_msg()
        self.static_transform_stamped_1.header.frame_id = "bumperbot_top"
        self.static_transform_stamped_1.child_frame_id = "bumperbot_adjacent"
        self.static_transform_stamped_1.transform.translation.x = 0.3
        self.static_transform_stamped_1.transform.translation.y = 0.0
        self.static_transform_stamped_1.transform.translation.z = 0.0
        self.static_transform_stamped_1.transform.rotation.x = 0.0
        self.static_transform_stamped_1.transform.rotation.y = 0.0
        self.static_transform_stamped_1.transform.rotation.z = 0.0
        self.static_transform_stamped_1.transform.rotation.w = 1.0

        self.static_tf_broadcaster.sendTransform([self.static_transform_stamped, self.static_transform_stamped_1])

        self.get_logger().info(f"Static transform between {self.static_transform_stamped.header.frame_id} and {self.static_transform_stamped.child_frame_id} published.")


        self.timer_ = self.create_timer(0.1, self.timer_callback)

        self.get_transform_service = self.create_service(GetTransform, 'get_transform', self.get_transform_callback)

    def timer_callback(self):
        self.dynamic_transform_stamped.header.stamp = self.get_clock().now().to_msg()        
        self.dynamic_transform_stamped.header.frame_id = "odom"
        self.dynamic_transform_stamped.child_frame_id = "bumperbot_base"
        self.dynamic_transform_stamped.transform.translation.x = self.last_x_ + self.x_increment_
        self.dynamic_transform_stamped.transform.translation.y = 0.0
        self.dynamic_transform_stamped.transform.translation.z = 0.0
        self.dynamic_transform_stamped.transform.rotation.x = 0.0
        self.dynamic_transform_stamped.transform.rotation.y = 0.0
        self.dynamic_transform_stamped.transform.rotation.z = 0.0
        self.dynamic_transform_stamped.transform.rotation.w = 1.0

        self.dynamic_tf_broadcaster.sendTransform(self.dynamic_transform_stamped)

        self.last_x_ = self.dynamic_transform_stamped.transform.translation.x

    def get_transform_callback(self, request, response):
        self.get_logger().info(f"Transform request received from {request.frame_id} to {request.child_frame_id}")
        requested_transform = TransformStamped()
        try:
            requested_transform = self.tf_buffer_.lookup_transform(request.frame_id, request.child_frame_id, rclpy.time.Time())
        except TransformException as ex:
            self.get_logger().error(f"Transform error between {request.frame_id} and {request.child_frame_id}: {ex}")
            response.success = False
            return response
        response.transform = requested_transform
        response.success = True
        return response
    
def main():
    rclpy.init()
    node = SimpleTFKinematics()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/bumperbot_cpp_examples/src/simple_tf_kinematics.cpp
#include <bumperbot_cpp_examples/simple_tf_kinematics.hpp>

using namespace std::chrono_literals;
using namespace std::placeholders;

SimpleTfKinematics::SimpleTfKinematics(const std::string &node_name) : Node(node_name), x_increment_(0.01), last_x_(0.0)
{
    static_tf_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
    dynamic_tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);

    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
    
    get_transform_service_ = create_service<bumperbot_msgs::srv::GetTransform>(
        "get_transform",
        std::bind(&SimpleTfKinematics::getTransformCallback, this, _1, _2));

    static_transform_stamped_.header.frame_id = "bumperbot_base";
    static_transform_stamped_.child_frame_id = "bumperbot_top";        
    static_transform_stamped_.header.stamp = get_clock()->now();
    static_transform_stamped_.transform.translation.x = 0.0;
    static_transform_stamped_.transform.translation.y = 0.0;
    static_transform_stamped_.transform.translation.z = 0.3;
    static_transform_stamped_.transform.rotation.x = 0.0;
    static_transform_stamped_.transform.rotation.y = 0.0;
    static_transform_stamped_.transform.rotation.z = 0.0;
    static_transform_stamped_.transform.rotation.w = 1.0;
    
    static_tf_broadcaster_->sendTransform(static_transform_stamped_);

    RCLCPP_INFO(this->get_logger(), "Static transform from base_link to base_footprint published.");

    dynamic_tf_timer_ = create_wall_timer(
        0.1s,
        std::bind(&SimpleTfKinematics::publishDynamicTransform, this));
}

void SimpleTfKinematics::publishDynamicTransform()
{
    dynamic_transform_stamped_.header.frame_id = "odom";
    dynamic_transform_stamped_.child_frame_id = "bumperbot_base";
    dynamic_transform_stamped_.header.stamp = get_clock()->now();
    dynamic_transform_stamped_.transform.translation.x = last_x_ + x_increment_;
    dynamic_transform_stamped_.transform.translation.y = 0.0;
    dynamic_transform_stamped_.transform.translation.z = 0.0;
    dynamic_transform_stamped_.transform.rotation.x = 0.0;
    dynamic_transform_stamped_.transform.rotation.y = 0.0;
    dynamic_transform_stamped_.transform.rotation.z = 0.0;
    dynamic_transform_stamped_.transform.rotation.w = 1.0;

    dynamic_tf_broadcaster_->sendTransform(dynamic_transform_stamped_);
    
    last_x_ += x_increment_;
}

bool SimpleTfKinematics::getTransformCallback(
    const std::shared_ptr<bumperbot_msgs::srv::GetTransform::Request> request,
    const std::shared_ptr<bumperbot_msgs::srv::GetTransform::Response> response)
{
    RCLCPP_INFO_STREAM(rclcpp::get_logger("SimpleTfKinematics"), "Received request for transform from " << request->frame_id.c_str() << " to " << request->child_frame_id.c_str());
    
    try {
        geometry_msgs::msg::TransformStamped transform = tf_buffer_->lookupTransform(request->frame_id, request->child_frame_id, tf2::TimePointZero);
        response->transform = transform;
        response->success = true;
        return true;
    } catch (tf2::TransformException &ex) {
        RCLCPP_ERROR_STREAM(rclcpp::get_logger("SimpleTfKinematics"), "Could not get transform: from " << request->frame_id.c_str() << " to " << request->child_frame_id.c_str() << " : " << ex.what());
        return false;
    }
}

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleTfKinematics>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}